storage/remote/queue_manager.go

```diff
@@ -763,11 +768,31 @@ func (t *QueueManager) reshardLoop() {
 	for {
 		select {
 		case numShards := <-t.reshardChan:
-			// We start the newShards after we have stopped (the therefore completely
-			// flushed) the oldShards, to guarantee we only every deliver samples in
-			// order.
-			t.shards.stop()
-			t.shards.start(numShards)
+			if len(t.shards.queues) > 0 {
+				// Send reshard signal only if old shards are exising.
+				t.shards.reshard <- struct{}{}
+			}
+			// We start the new shards parallel to the old shards. These new shards
+			// are filled with existing samples in old shards while blocking the
+			// incoming samples through Append. Filling of existing samples is pretty
+			// quick, so loss in time is negligible.
+			//
+			// Once the existing samples are filled, the old shards are replaced by
+			// the new shards and then incoming samples are consumed. This is done
+			// to guarantee that we deliver samples in order only.
+			// Create new shards.
+			newShards := t.newShards()
+			newShards.start(numShards)
+			// Block incoming samples and wait until the exxisting samples are transfered
+			// to the newly created shards.
+			t.blockIncomingSamples.Store(true)
+			t.shards.stop(true)
+			t.shards.transferToShards(newShards)
+			// Resume accepting samples after the old samples have been transfered.
+			// This ensures that samples are sorted according to time.
+			t.blockIncomingSamples.Store(false)
```

Comment:
Shouldn't sends be blocked until the new shards replace the old shards (two lines down)?